src/Optimization/hiopAlgPrimalDecomp.cpp

```diff
@@ -0,0 +1,1211 @@
+//include header file
+#include "hiopAlgPrimalDecomp.hpp"
+#include "hiopInterfacePrimalDecomp.hpp"
+
+#include <cstring>
+#include <cassert>
+
+namespace hiop
+{
+#ifdef HIOP_USE_MPI
+  /* This struct provides the info necessary for the recourse approximation function
+   * buffer[n+1] contains both the function value and gradient w.r.t x
+   * buffer[0] is the function value and buffer[1:n] the gradient
+   * Contains send and receive functionalities for the values in buffer
+   */
+  struct ReqRecourseApprox
+  {
+    ReqRecourseApprox() : ReqRecourseApprox(1) {}
+    ReqRecourseApprox(const int& n)
+    {
+      n_=n;
+      buffer=new double[n_+1];
+    }
+
+    int test() 
+    {
+      int mpi_test_flag; MPI_Status mpi_status;
+      int ierr = MPI_Test(&request_, &mpi_test_flag, &mpi_status);
+      assert(MPI_SUCCESS == ierr);
+      return mpi_test_flag;
+    }
+    void post_recv(int tag, int rank_from, MPI_Comm comm)
+    {
+      int ierr = MPI_Irecv(buffer, n_+1, MPI_DOUBLE, rank_from, tag, comm, &request_);
+      assert(MPI_SUCCESS == ierr);
+    }
+    void post_send(int tag, int rank_to, MPI_Comm comm)
+    {
+      int ierr = MPI_Isend(buffer, n_+1, MPI_DOUBLE, rank_to, tag, comm, &request_);
+      assert(MPI_SUCCESS == ierr);
+    }
+    double value(){return buffer[0];}
+    void set_value(const double v){buffer[0]=v;}
+    double grad(int i){return buffer[i+1];}
+    void set_grad(const double* g)
+    {
+      for(int i=0;i<n_;i++) {
+        buffer[i+1]=g[i];
+      }
+    }
+
+    MPI_Request request_;
+  private:
+    int n_;
+    double* buffer;
+  };
+
+  /* This struct is used to post receive and request for contingency
+   * index that is to be solved by the solver ranks.
+   */
+  struct ReqContingencyIdx
+  {
+    ReqContingencyIdx() : ReqContingencyIdx(-1) {}
+    ReqContingencyIdx(const int& idx_)
+    {
+      idx=idx_;
+    }
+
+    int test() {
+      int mpi_test_flag; MPI_Status mpi_status;
+      int ierr = MPI_Test(&request_, &mpi_test_flag, &mpi_status);
+      assert(MPI_SUCCESS == ierr);
+      return mpi_test_flag;
+    }
+    void post_recv(int tag, int rank_from, MPI_Comm comm)
+    {
+      int ierr = MPI_Irecv(&idx, 1, MPI_INT, rank_from, tag, comm, &request_);
+      assert(MPI_SUCCESS == ierr);
+    }
+    void post_send(int tag, int rank_to, MPI_Comm comm)
+    {
+      int ierr = MPI_Isend(&idx, 1, MPI_INT, rank_to, tag, comm, &request_);
+      assert(MPI_SUCCESS == ierr);
+    }
+    double value(){return idx;}
+    void set_idx(const int& i){idx = i;}
+    MPI_Request request_;
+  private:
+    int idx;
+  };
+#endif
+
+
+
+hiopAlgPrimalDecomposition::HessianApprox::
+HessianApprox() : 
```

Comment:
I would avoid using nested classes. Better have a standalone implementation of Hessian approximation and then use composition with `hiopAlgPrimalDecomposition` class.

Reply:
Hoping to discuss this issue later this week.

Reply:
We agreed the nested class approach is appropriate here.